File: app/desktop/src/lib/types/kicad.ts
/**
 * KiCad IPC contracts: connection state, write results, live board
 * diagnostics, write preconditions, and dry-run coil previews.
 *
 * Mirrors the Rust `BoardDiagnosticsIpc` / `CoilPreviewIpc` wire formats
 * (snake_case); Tauri's IPC layer converts to/from camelCase on the JS
 * side automatically, so the field names here match the wire format.
 */

export interface KicadConnection {
  connected: boolean;
  board_name: string;
  copper_layers: number;
}

export interface KicadWriteResult {
  /** Number of items we sent to KiCad (tracks + vias). */
  items_attempted: number;
  /** Number of items KiCad accepted (ItemStatus.code == ISC_OK). */
  items_created: number;
  /**
   * Up to 1000 per-item failure messages from KiCad. Empty when
   * `items_created === items_attempted`. The full count of failures is
   * always `items_attempted - items_created` even if only a subset are
   * listed here.
   */
  failures: string[];
  /**
   * Summary of all rejection codes from KiCad, sorted by count descending.
   * Each entry is `[code, count]` where code is the ItemStatusCode
   * (1=OK, 2=invalid type, 3=existing, 4=non-existent, 5=immutable,
   * 7=invalid data). Empty when all items succeeded.
   */
  failure_summary: [number, number][];
  /**
   * Commit ID shown in KiCad's undo stack. `"atomic-commit"` on a real
   * write, `"(dry run - no commit)"` when written with `dry_run: true`.
   */
  commit_id: string;
}

export interface KicadPingResult {
  ok: boolean;
  version: string;
}

/** Live snapshot of the open KiCad board (get_board_diagnostics). */
export interface BoardDiagnostics {
  /** File name of the open board, e.g. `"board.kicad_pcb"`. */
  board_name: string;
  /** Number of copper layers enabled on the board. */
  copper_layer_count: number;
  /** Bounding box of the board's edge cuts [mm]. 0 if not queryable. */
  board_x_min_mm: number;
  /** See board_x_min_mm. */
  board_x_max_mm: number;
  /** See board_x_min_mm. */
  board_y_min_mm: number;
  /** See board_x_min_mm. */
  board_y_max_mm: number;
  /** Net class names defined on the board. Empty if not queryable. */
  available_net_classes: string[];
}

/** Severity level for pre-condition warnings. */
export type PreconditionLevel = "info" | "warning" | "error";

/**
 * One warning / recommendation about the (config, board) pair, produced by
 * `validate_write_preconditions`. The UI renders `message` verbatim and
 * colour-codes by `level`; `field` optionally highlights the offending input.
 */
export interface PreconditionWarning {
  level: PreconditionLevel;
  field: string | null;
  message: string;
}

/** Per-layer breakdown of the coils that would be written. */
export interface CoilPreviewLayer {
  layer_idx: number;
  /** Number of phase coils on this layer. */
  phase_count: number;
  /** Total track segments (sum of `segments.length` across phases). */
  segment_count: number;
  /** Inter-layer vias on this layer. */
  via_count: number;
}

/**
 * Dry-run summary of what `write_coils_to_board` would produce (preview_coils).
 * Precondition warnings are not included — the UI calls
 * `validateWritePreconditions` separately.
 */
export interface CoilPreview {
  /** Number of layers the writer would iterate over. */
  num_layers: number;
  /** Routing-pattern id used for this preview, e.g. `"infinity-braid"`. */
  topology: string;
  /** Per-layer breakdown. */
  layers: CoilPreviewLayer[];
  /** Total track segments across all layers. */
  total_tracks: number;
  /** Total vias across all layers. */
  total_vias: number;
}

File: app/desktop/src-tauri/src/ipc/kicad.rs
//! IPC DTOs for the KiCad bridge: board diagnostics, write preconditions, and
//! coil preview, with their core `pcbmotorgen_export` converters.

use serde::{Deserialize, Serialize};

use super::enums::PreconditionLevelIpc;

// ===========================================================================
// Board diagnostics (get_board_diagnostics / validate_write_preconditions)
// ===========================================================================

/// Snapshot of the open KiCad board — IPC wire format.
///
/// Mirrors `pcbmotorgen_export::BoardDiagnostics` exactly. `board_*_mm`
/// are populated from the board's edge cuts when the IPC supports that query
/// (currently 0.0 — TODO: real query). `available_net_classes` is empty for
/// the same reason. `board_name` and `copper_layer_count` are always
/// populated by `get_board_diagnostics`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct BoardDiagnosticsIpc {
    pub board_name: String,
    pub copper_layer_count: u32,
    /// Bounding box of the board edge cuts, in mm. Defaults to 0.0 when
    /// the KiCad 10 IPC does not yet expose the edge-cut query (TODO).
    pub board_x_min_mm: f64,
    pub board_x_max_mm: f64,
    pub board_y_min_mm: f64,
    pub board_y_max_mm: f64,
    pub available_net_classes: Vec<String>,
}

impl BoardDiagnosticsIpc {
    /// Convert a core `BoardDiagnostics` to the IPC wire format.
    pub fn from_core(b: &pcbmotorgen_export::BoardDiagnostics) -> Self {
        Self {
            board_name: b.board_name.clone(),
            copper_layer_count: b.copper_layer_count,
            board_x_min_mm: b.board_x_min_mm,
            board_x_max_mm: b.board_x_max_mm,
            board_y_min_mm: b.board_y_min_mm,
            board_y_max_mm: b.board_y_max_mm,
            available_net_classes: b.available_net_classes.clone(),
        }
    }
}

/// One warning or recommendation about the (config, board) pair.
///
/// Produced by `validate_write_preconditions`. The UI is expected to render
/// `message` verbatim and colour-code by `level`. `field` is an optional
/// machine-readable key (`"num_layers"`, `"active_area_length_m"`, …) the
/// UI can use to highlight the offending input control.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct PreconditionWarningIpc {
    pub level: PreconditionLevelIpc,
    pub field: Option<String>,
    pub message: String,
}

impl PreconditionWarningIpc {
    /// Convert a core `PreconditionWarning` to the IPC wire format.
    pub fn from_core(w: &pcbmotorgen_export::PreconditionWarning) -> Self {
        let level = match w.level {
            pcbmotorgen_export::PreconditionLevel::Info => PreconditionLevelIpc::Info,
            pcbmotorgen_export::PreconditionLevel::Warning => {
                PreconditionLevelIpc::Warning
            }
            pcbmotorgen_export::PreconditionLevel::Error => PreconditionLevelIpc::Error,
        };
        Self {
            level,
            field: w.field.clone(),
            message: w.message.clone(),
        }
    }
}

// ===========================================================================
// Coil preview (preview_coils)
// ===========================================================================

/// Per-layer breakdown of the coils that would be written.
///
/// Mirrors `pcbmotorgen_export::CoilPreviewLayer` minus `board_layer` —
/// the UI infers the layer assignment from `layer_idx` (the writer maps
/// `layer_idx == 0` → B.Cu, `layer_idx == num_layers-1` → F.Cu).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct CoilPreviewLayerIpc {
    pub layer_idx: u32,
    pub phase_count: u32,
    pub segment_count: u32,
    pub via_count: u32,
}

/// Dry-run summary of what `write_coils_to_board` would produce.
///
/// Returned by `preview_coils`. Contains the per-layer tally and the
/// topology label. Pre-condition warnings are *not* included here — the
/// UI calls `validate_write_preconditions` separately for those. The full
/// `PhaseCoil` geometry is *not* carried on the wire here either — the
/// UI calls `generate_coils` separately if it needs the raw segments.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct CoilPreviewIpc {
    pub num_layers: u32,
    /// Topology label — `"serpentine" | "sine_wave" | "concentrated" |
    /// "rhombic" | "spiral"`. Matches the core's `topology_label()` output.
    pub topology: String,
    pub layers: Vec<CoilPreviewLayerIpc>,
    pub total_tracks: u32,
    pub total_vias: u32,
}

impl CoilPreviewIpc {
    /// Convert a core `CoilPreview` to the IPC wire format.
    ///
    /// Note: `p.topology` is already a `String` (set by the core's
    /// `topology_label()`), so we just clone it — no enum match needed.
    /// The core's `CoilPreview` does not carry a `warnings` field; the UI
    /// calls `validate_write_preconditions` separately for those.
    pub fn from_core(p: &pcbmotorgen_export::CoilPreview) -> Self {
        let layers = p
            .layers
            .iter()
            .map(|l| CoilPreviewLayerIpc {
                layer_idx: l.layer_idx,
                phase_count: l.phase_count,
                segment_count: l.segment_count,
                via_count: l.via_count,
            })
            .collect();
        Self {
            num_layers: p.num_layers,
            topology: p.topology.clone(),
            layers,
            total_tracks: p.total_tracks,
            total_vias: p.total_vias,
        }
    }
}